packages/playwright-core/src/server/browserContext.ts

```diff
@@ -513,17 +514,88 @@ export abstract class BrowserContext extends SdkObject {
     };
     const originsToSave = new Set(this._origins);
 
+    async function _collectStorageScript() {
+
+      const idbResult = await Promise.all((await indexedDB.databases()).map(async dbInfo => {
+        if (!dbInfo.name)
+          throw new Error('Database name is empty');
+
+        function idbRequestToPromise<T extends IDBOpenDBRequest | IDBRequest>(request: T) {
+          return new Promise<T['result']>((resolve, reject) => {
+            request.addEventListener('success', () => resolve(request.result));
+            request.addEventListener('error', () => reject(request.error));
+          });
+        }
+
+        const db = await idbRequestToPromise(indexedDB.open(dbInfo.name));
+        const transaction = db.transaction(db.objectStoreNames, 'readonly');
+        const stores = await Promise.all([...db.objectStoreNames].map(async storeName => {
+          const objectStore = transaction.objectStore(storeName);
+
+          const keys = await idbRequestToPromise(objectStore.getAllKeys());
+          const records = await Promise.all(keys.map(async key => {
+            return {
+              key: objectStore.keyPath === null ? key : undefined,
+              value: await idbRequestToPromise(objectStore.get(key))
+            };
+          }));
+
+          const indexes = [...objectStore.indexNames].map(indexName => {
+            const index = objectStore.index(indexName);
+            return {
+              name: index.name,
+              keyPath: typeof index.keyPath === 'string' ? index.keyPath : undefined,
+              keyPathArray: Array.isArray(index.keyPath) ? index.keyPath : undefined,
+              multiEntry: index.multiEntry,
+              unique: index.unique,
+            };
+          });
+
+          return {
+            name: storeName,
+            records: records.filter(Boolean),
+            indexes,
+            autoIncrement: objectStore.autoIncrement,
+            keyPath: typeof objectStore.keyPath === 'string' ? objectStore.keyPath : undefined,
+            keyPathArray: Array.isArray(objectStore.keyPath) ? objectStore.keyPath : undefined,
+          };
+        }));
+
+        return {
+          name: dbInfo.name,
+          version: dbInfo.version,
+          stores,
+        };
+      }));
+
+      return {
+        localStorage: Object.keys(localStorage).map(name => ({ name, value: localStorage.getItem(name) })),
+        indexedDB: idbResult,
+      };
+    }
+
+    function serializeRecords(indexedDBs: channels.IndexedDBDatabase[]) {
+      for (const db of indexedDBs) {
+        for (const store of db.stores) {
+          for (const record of store.records) {
+            if (record.key !== undefined)
+              record.key = JSON.stringify(utilitySerializers.serializeAsCallArgument(record.value, v => ({ fallThrough: v })));
```

Comment:
WDYT about including keys/values verbatim, as long as they are serializable as JSON? And if not, we can have additional `keyEncoded`/`valueEncoded` properties that will hold the serialized value in any format. This way, the format is usually readable, except for when encoding is actually required.

Reply:
makes sense to me 👍 doing this in a follow-up